lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -0,0 +1,2152 @@
+                   .attributes = getMandatoryPortAttr(ctx, "done")});
+}
+
+/// Returns true if @p groupOp is stateful, i.e. if there is some
+/// calyx.GroupDoneOp within the group.
+static bool isStatefulGroup(calyx::GroupOp groupOp) {
+  return !groupOp.getOps<calyx::GroupDoneOp>().empty();
+}
+
+/// Creates a new group within @p compOp.
+template <typename TGroup, typename TRet = TGroup>
+static TRet createGroup(PatternRewriter &rewriter, calyx::ComponentOp compOp,
+                        Location loc, Twine uniqueName) {
+  return persistInsertionPoint(rewriter, [&] {
+    rewriter.setInsertionPointToEnd(compOp.getWiresOp().getBody());
+    auto groupOp = rewriter.create<TGroup>(loc, uniqueName.str());
+    rewriter.createBlock(&groupOp.getBodyRegion());
+    return groupOp;
+  });
+}
+
+/// ComponentLoweringState handles the current state of lowering of a calyx
+/// component. It is mainly used as a key/value store for recording information
+/// during partial lowering, which is required at later lowering passes.
+class ProgramLoweringState;
+class ComponentLoweringState {
+public:
+  ComponentLoweringState(ProgramLoweringState &pls, calyx::ComponentOp compOp)
+      : programLoweringState(pls), compOp(compOp) {}
+
+  ProgramLoweringState &getProgramState() { return programLoweringState; }
+
+  /// Returns a unique name within @p compOp with the provided @p prefix.
+  std::string getUniqueName(Twine prefix) {
+    return (prefix + "_" + std::to_string(prefixIdMap[prefix.str()]++)).str();
+  }
+
+  /// Returns a unique name associated with a specific operation.
+  StringRef getUniqueName(Operation *op) {
+    auto it = opNames.find(op);
+    assert(it != opNames.end() && "A unique name should have been set for op");
+    return it->second;
+  }
+
+  /// Registers a unique name for a given operation using a provided prefix.
+  void setUniqueName(Operation *op, Twine prefix) {
+    auto it = opNames.find(op);
+    assert(it == opNames.end() && "A unique name was already set for op");
+    std::string un = getUniqueName(prefix);
+    opNames[op] = un;
+  }
+
+  /// Returns the calyx::ComponentOp associated with this lowering state.
+  calyx::ComponentOp getCompOp() { return compOp; }
+
+  /// Returns a unique name for instantiating a new instance of component @p
+  /// comp within this component.
+  std::string genNewInstanceName(calyx::ComponentOp comp) {
+    int i = 0;
+    compOp.walk([&](calyx::InstanceOp instOp) {
+      if (instOp.getReferencedComponent() == comp)
+        i++;
+    });
+    return comp.getName().str() + "_inst_" + std::to_string(i);
+  }
+
+  /// Returns an ordered list of schedulables which registerred themselves to be
+  /// a result of lowering the @p block in the source program. The list order
+  /// follows def-use chains between the groups.
+  SmallVectorImpl<Scheduleable> &getBlockScheduleables(mlir::Block *block) {
+    return blockScheduleables[block];
```

Comment:
Blocks may have no `Scheduleable`s, i.e., if they consisted of purely combinational logic. However, for clarity, i'll change it to returning an empty vector instead of the default constructed value of the map.